File: impl-go/no_0009_palindrome_number/no_0009_palindrome_number.go
package no_0009_palindrome_number

func isPalindrome(x int) bool {
	if x < 0 || (x != 0 && x%10 == 0) {
		return false
	}

	n := 0
	for x > n {
		n = n*10 + x%10 // 把 x 的最后一位追加到 n 的后面
		x /= 10
	}

	return n == x || n/10 == x
}

File: impl-rust/src/no_0009_palindrome_number.rs
pub fn is_palindrome(x: i32) -> bool {
    if x < 0 || (x != 0 && x % 10 == 0) {
        return false;
    }

    let mut n = 0;
    let mut x = x;
    while x > n {
        n = n * 10 + x % 10;
        x /= 10;
    }

    x == n || n / 10 == x
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_is_palindrome() {
        assert_eq!(is_palindrome(121), true);
        assert_eq!(is_palindrome(-121), false);
        assert_eq!(is_palindrome(10), false);
        assert_eq!(is_palindrome(0), true);
    }
}
